layout: Add line height from preserved segment breaks in quirks mode

mrobinson opened the pull request:
In quirks mode, preserved segment breaks should add line height to
lines. This matches the behavior of WebKit and Blink, but not Gecko.

This also handles the special-case of `<br>` elements, which are
implemented with preserved segment breaks via `white-space: pre-line`.
This is an implementation detail though because `<br>` has a special
behavior if the line isn't empty -- it doesn't add any line height in
this case.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes OR


<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

mrobinson committed:
layout: Add line height from preserved segment breaks in quirks mode

In quirks mode, preserved segment breaks should add line height to
lines. This matches the behavior of WebKit and Blink, but not Gecko.

This also handles the special-case of `<br>` elements, which are
implemented with preserved segment breaks via `white-space: pre-line`.
This is an implementation detail though because `<br>` has a special
behavior if the line isn't empty -- it doesn't add any line height in
this case.

mrobinson commented:
This needs a bit more work. In quirks mode forced line breaks do not add line-height if the line is otherwise empty, but `<br>` elements do! I think it is finally the moment to add "magic" behavior for `<br>`.

Loirooriol reviewed (changes requested):
The spec says

> It either contains no text or it contains only collapsed whitespace.

So a forced line break doesn't fulfill the condition, right? I'm testing:

```html
a
<div style="white-space: pre">&#xa;&#xa;&#xa;</div>
b
<div style="white-space: pre"><span>&#xa;&#xa;&#xa;</span></div>
c
<div><span style="white-space: pre">&#xa;&#xa;&#xa;</span></div>
d
```

I think Blink and WebKit are correct: all letters are separated.
Firefox seems to have a bug: c and d are together.
And Servo is worse: all letters are together.

So I think:
 - We shouldn't add `<br>` magic here.
 - We align Servo with Blink/WebKit.
 - We should add a WPT that fails in Firefox.

mrobinson commented:
There is still going to be some special behavior though, because `<br>` doesn't always add its line-height to the line. See how only the `<br>` on the empty line adds the full font size to line height in the following quirks mode examples:

```html
<!-- No br on empty line, so lines render as if font-size: 20px -->
<div style="font-size: 100px; border: solid;">
    <span style="font-size: 20px">Hello</span><br>
    <span style="font-size: 20px">Hello2</span><br>
</div>

<!-- br on empty line, so middle line renders as if font-size: 100px -->
<div style="font-size: 100px; border: solid;">
    <span style="font-size: 20px">Hello</span><br>
    <br>
    <span style="font-size: 20px">Hello2</span><br>
</div>

<!-- The same as the first example, this renders differently than with br in all browsers -->
<div style="font-size: 100px; border: solid; white-space: pre-line"><span style="font-size: 20px">Hello</span><br>
    <span style="font-size: 20px">Hello2</span><br>
</div>

```

mrobinson commented:
Okay. I've modified the change to always add the strut size for preserved segment breaks, except in the case for `<br>` which shouldn't do this when the line isn't empty. I've also added a test to verify this for all browsers. This test fails only in Gecko.

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#44785) with upstreamable changes.

mrobinson commented:
> Possibly may be left for a follow-up, but note this differs from Gecko/Blink/WebKit:
> 
> ```
> a<div style="font-size: 0">x<span style="font-size: 16px"> </span>x</div>b
> ```

Oh yeah, that's an issue. I think it's covered by this todo from `inline.rs`:

```rust
            // TODO(mrobinson): This isn't quite right for situations where collapsible white space
            // ultimately does not collapse because it is between two other pieces of content.
```

Loirooriol commented on tests/wpt/tests/quirks/line-height-preserved-segment-break.html:
```diff
@@ -0,0 +1,27 @@
+
+<!-- This div is just to make sure that red shows up if the test div doesn't
+     display properly. -->
+<div style="position: absolute; z-index: -100; width: 100px; height: 100px; background: red;"></div>
```
Fine either way, but you could turn this into a wrapper to avoid `position: absolute; z-index: -100`:

```html
<div style="width: 100px; height: 100px; background: red;">
  <div class="container"><span class="inline-box">&#xA;</span></div>
</div>
```

I also think the HTML looks slightly visually clearer with `&#xA;` rather than starting one line with `<div>` and next line with `</span>`. But again, not a big deal.

Loirooriol commented on tests/wpt/tests/quirks/line-height-preserved-segment-break.html:
```diff
@@ -0,0 +1,27 @@
+<link rel="author" title="Martin Robinson" href="mrobinson@igalia.com">
+<link rel="help" href="https://quirks.spec.whatwg.org/#the-line-height-calculation-quirk">
+<link rel="help" href="https://drafts.csswg.org/css-text/#valdef-white-space-pre-line">
+<link rel="match" href="line-height-preserved-segment-break-ref.html">
```
Maybe use `/css/reference/ref-filled-green-100px-square.xht`? Or do you want the ref to be inside `/quirks/`?

Loirooriol commented on tests/wpt/tests/quirks/line-height-preserved-segment-break.html:
```diff
@@ -0,0 +1,27 @@
+<link rel="match" href="line-height-preserved-segment-break-ref.html">
+<p>Test passes if there is a filled green square and <strong>no red</strong>.</p>
+
+<style>
```
You can move the `<style>` above so that it goes into the `<head>` instead of the `<body>`, which is invalid HTML.

mrobinson commented:
Thanks for the review, @Loirooriol. I've made the changes you suggested.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#44785).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#44785).
